src/clickonce/launcher/HostFinder.cs

```diff
@@ -0,0 +1,126 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using Microsoft.Deployment.Utilities;
+
+namespace Microsoft.Deployment.Launcher
+{
+    internal static class HostFinder
+    {
+        /// <summary>
+        /// Gets full path to .NET host appropriate for activating the application.
+        /// 
+        /// .NET host'slocation can be obtained from multiple locations.
+        ///
+        /// Current code searches in default, global shared runtime, location only:
+        /// %ProgramFiles%\dotnet and %ProgramFiles(x86)%\dotnet
+        /// 
+        /// Consider adding support for other non-standard registrations of host location.
+        /// 1) Environment variables: DOTNET_ROOT or DOTNET_ROOT(x86)
+        /// 2) Registry: HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\{arch}\[InstallLocation]
+        /// 
+        /// Order of search should eventually be:
+        /// 1) Environment variables
+        /// 2) Registry
+        /// 3) Global location
+        /// </summary>
+        /// <param name="applicationFilePath">Full path to application</param>
+        /// <returns></returns>
+        internal static string GetHost(string applicationFilePath)
+        {
+            string arch = GetProcessorArchitectureFromAssembly(applicationFilePath);
+
+            if (arch == "x86")
+            {
+                return GetX86Host();
+            }
+            else if (arch == "amd64" || arch == "arm64")
+            {
+                return Get64bitHost();
+            }
+            else if (arch == "msil")
+            {
+                string host = Get64bitHost();
+                return string.IsNullOrEmpty(host) ? GetX86Host() : host;
+            }
+
+            return string.Empty;
+        }
+
+        /// <summary>
+        /// Gets ProgramFiles folder for the specified bitness.
+        /// </summary>
+        /// <param name="is64bit">If 64-bit bitness is required</param>
+        /// <returns></returns>
+        private static string GetProgramFilesFolder(bool is64bit = false)
+        {
+            return is64bit ?
+                (Environment.Is64BitProcess ?
+                    Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles) :
+                    Environment.GetEnvironmentVariable("ProgramW6432")) :
+                Environment.GetFolderPath(Environment.SpecialFolder.ProgramFilesX86);
+        }
+
+        /// <summary>
+        /// Get global host if it exists, for the specified bitness.
+        /// </summary>
+        /// <param name="is64bit">If 64-bit bitness is required</param>
+        /// <returns></returns>
+        private static string GetGlobalHost(bool is64bit = false)
+        {
+            string folder = GetProgramFilesFolder(is64bit);
+            string host = !string.IsNullOrEmpty(folder) ? Path.Combine(folder, "dotnet\\dotnet.exe") : string.Empty;
```

Comment:
Why would `folder` ever be `null`? Can ClickOnce work at all if Program Files can't be found?

Reply:
Yes - ClickOnce doesn't need anything from Program Files. It uses .NET FX runtime and persist applications under %userprofile%